File: 394. Decode String.py
class Solution(object):
    def decodeString(self, s):
        """
        :type s: str
        :rtype: str
        >>> s = Solution()
        >>> s.decodeString('fa3[b2[d]]e')
        'fabddbddbdde'
        """
        repeat_starts, ret, times = [], '', 0
        for c in s:
            if c == '[':
                repeat_starts.append((times, len(ret)))
                times = 0
            elif c == ']':
                pair = repeat_starts.pop()
                ret += (pair[0] - 1) * ret[pair[1] :]
            elif c.isdigit():
                times = times * 10 + int(c)
            else:
                ret += c
        return ret

if __name__ == "__main__":
    import doctest
    doctest.testmod()

File: 394. Decode String.cpp
class Solution {
public:
    string decodeString(string s) {
        stack<pair<int, int>> repeat_starts;
        string ret;
        int times = 0;
        for(auto& c : s) {
          if(c == '[') {
            repeat_starts.push(make_pair(times - 1, ret.size()));
            times = 0;
          }
          else if(c == ']') {
            auto repeat_info = repeat_starts.top();
            auto repeat_content = ret.substr(repeat_info.second);
            for(int k = 1; k <= repeat_info.first; ++k) {
              ret.append(repeat_content);
            }
            repeat_starts.pop();
          }
          else if(isdigit(c)){
            times = times * 10 + c - '0';
          }
          else {
            ret.push_back(c);
          }
        }
        return ret;
    }
};
